pkg/workflow/providers/multicluster/deploy.go

```diff
@@ -148,80 +164,230 @@ func overrideConfiguration(policies []v1beta1.AppPolicy, components []common.App
+func (t *applyTask) allInputFilled(cache *pkgsync.Map[string, *value.Value]) bool {
+	for _, in := range t.component.Inputs {
+		if val := t.getVar(in.From, cache); val == nil {
+			return false
+		}
 	}
-	return dependents
+
+	return true
 }
 
 type applyTaskResult struct {
 	healthy bool
 	err     error
+	task    *applyTask
 }
 
+// applyComponents will apply components to placements.
 func applyComponents(ctx context.Context, apply oamProvider.ComponentApply, healthCheck oamProvider.ComponentHealthCheck, components []common.ApplicationComponent, placements []v1alpha1.PlacementDecision, parallelism int) (bool, string, error) {
 	var tasks []*applyTask
+	var cache = pkgsync.NewMap[string, *value.Value]()
+	rootValue, err := value.NewValue("{}", nil, "")
+	if err != nil {
+		return false, "", err
+	}
+	var cueMutex sync.Mutex
+	var makeValue = func(s string) (*value.Value, error) {
+		cueMutex.Lock()
+		defer cueMutex.Unlock()
+		return rootValue.MakeValue(s)
+	}
+
+	taskHealthyMap := map[string]bool{}
 	for _, comp := range components {
 		for _, pl := range placements {
 			tasks = append(tasks, &applyTask{component: comp, placement: pl})
 		}
 	}
-	healthCheckResults := parallel.Run(func(task *applyTask) *applyTaskResult {
-		healthy, err := healthCheck(ctx, task.component, nil, task.placement.Cluster, task.placement.Namespace, "")
-		return &applyTaskResult{healthy: healthy, err: err}
-	}, tasks, parallelism).([]*applyTaskResult)
-	taskHealthyMap := map[string]bool{}
-	for i, res := range healthCheckResults {
-		taskHealthyMap[tasks[i].key()] = res.healthy
+	healthCheckError := make([]*applyTaskResult, 0)
+	maxHealthCheckTimes := len(tasks)
+HealthCheck:
+	for i := 0; i < maxHealthCheckTimes; i++ {
+		checkTasks := make([]*applyTask, 0)
+		for _, task := range tasks {
+			if task.healthy == nil && task.allDependsReady(taskHealthyMap) && task.allInputFilled(cache) {
+				task.healthy = new(bool)
+				err := task.fillInputs(cache, makeValue)
+				if err != nil {
+					taskHealthyMap[task.key()] = false
+					healthCheckError = append(healthCheckError, &applyTaskResult{healthy: false, err: err, task: task})
+					continue
+				}
+				checkTasks = append(checkTasks, task)
+			}
+		}
+		if len(checkTasks) == 0 {
+			break HealthCheck
+		}
+		checkResults := slices.ParMap[*applyTask, *applyTaskResult](checkTasks, func(task *applyTask) *applyTaskResult {
+			healthy, output, outputs, err := healthCheck(ctx, task.component, nil, task.placement.Cluster, task.placement.Namespace, "")
+			task.healthy = pointer.Bool(healthy)
+			if healthy {
+				err = task.generateOutput(output, outputs, cache, makeValue)
+			}
+			return &applyTaskResult{healthy: healthy, err: err, task: task}
+		}, slices.Parallelism(parallelism))
+
+		for _, res := range checkResults {
+			taskHealthyMap[res.task.key()] = res.healthy
+			if !res.healthy || res.err != nil {
```

Comment:
If is not healthy, the result will be put in healthCheckError? The variable name healthCheckError looks a bit confusing.

Reply:
It's a little confusing and wrong. Maybe we should call it `unhealthResult`?

Reply:
unhealthyResults looks better